File: vastaanotto/src/main/scala/fi/oph/viestinvalitys/vastaanotto/model/Lahetys.scala
package fi.oph.viestinvalitys.vastaanotto.model

import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.media.Schema.RequiredMode

import java.util.UUID
import scala.annotation.meta.field
import scala.beans.BeanProperty

object Lahetys {
  final val OTSIKKO_MAX_PITUUS = 255
}

/**
 * Lähetyksen otsikko
 *
 * @param otsikko
 */
case class Lahetys(
  @(Schema @field)(example = "Lähetyksen Otsikko", requiredMode=RequiredMode.REQUIRED, maxLength = Lahetys.OTSIKKO_MAX_PITUUS)
  @BeanProperty otsikko: String,

  @(Schema@field)(requiredMode = RequiredMode.REQUIRED, example = "[\"APP_ATARU_HAKEMUS_CRUD_1.2.246.562.00.00000000000000006666\"]")
  @BeanProperty kayttooikeusRajoitukset: java.util.List[String],
                  ) {
  /**
   * Tyhjä konstruktori Jacksonia varten
   */
  def this() = {
    this(null, null)
  }
}

File: lahetysrajapinta/src/main/scala/fi/oph/viestinvalitys/vastaanotto/model/Lahetys.java
package fi.oph.viestinvalitys.vastaanotto.model;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import io.swagger.v3.oas.annotations.media.Schema;

import java.util.List;
import java.util.Optional;

public interface Lahetys {
  
  Optional<String> getOtsikko();

  Optional<String> getLahettavaPalvelu();

  Optional<String> getLahettavanVirkailijanOid();

  @JsonDeserialize(as = LahettajaImpl.class)
  @Schema(implementation = LahettajaImpl.class)
  public interface Lahettaja {

    Optional<String> getNimi();

    Optional<String> getSahkopostiOsoite();
  }

  Optional<Lahettaja> getLahettaja();

  Optional<String> getReplyTo();

  Optional<String> getPrioriteetti();

  Optional<Integer> getSailytysaika();


  interface OtsikkoBuilder {
    LahettavaPalveluBuilder withOtsikko(String otsikko);
  }

  interface LahettavaPalveluBuilder {
    LahettajaBuilder withLahettavaPalvelu(String nimi);
  }

  interface LahettajaBuilder {

    PrioriteettiBuilder withLahettaja(Optional<String> nimi, String sahkopostiOsoite);
  }

  interface PrioriteettiBuilder {

    SailytysaikaBuilder withNormaaliPrioriteetti();

    SailytysaikaBuilder withKorkeaPrioriteetti();
  }

  interface SailytysaikaBuilder {

    LahetysBuilder withSailytysaika(int sailytysaika);
  }

  interface LahetysBuilder {

    LahetysBuilder withReplyTo(String replyTo);
    LahetysBuilder withLahettavanVirkailijanOid(String oid);
    Lahetys build() throws BuilderException;
  }
}
